auth/token-managers/token-manager.ts

```diff
@@ -0,0 +1,240 @@
+/**
+ * Copyright 2020 IBM Corp. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { OutgoingHttpHeaders } from 'http';
+import logger from '../../lib/logger';
+import { RequestWrapper } from '../../lib/request-wrapper';
+import {getCurrentTime} from "../utils";
+
+/** Configuration options for JWT token retrieval. */
+export type TokenManagerOptions = {
+  /** The endpoint for token requests. */
+  url?: string;
+  /** Headers to be sent with every service token request. */
+  headers?: OutgoingHttpHeaders;
+  /**
+   * A flag that indicates whether verification of
+   *   the server's SSL certificate should be disabled or not.
+   */
+  disableSslVerification?: boolean;
+  /** Allow additional request config parameters */
+  [propName: string]: any;
+}
+
+/**
+ * A class for shared functionality for parsing, storing, and requesting
+ * JWT tokens. Intended to be used as a parent to be extended for token
+ * request management. Child classes should implement `requestToken()`
+ * to retrieve the bearer token from intended sources.
+ */
+export class TokenManager {
+  protected url: string;
+  protected tokenName: string;
+  protected disableSslVerification: boolean;
+  protected headers: OutgoingHttpHeaders;
+  protected requestWrapperInstance: RequestWrapper;
+  protected tokenInfo: any;
```

Comment:
While we're abstracting, I think we should delegate `tokenInfo` and `tokenName` to subclasses of `TokenManager`, with only `accessToken` visible at this level.

Reply:
Sorry, but I am not catching your point. `TokenInfo` currently in use to check that a token already is stored in the object. Here is a reference to a line where there properties are used:
https://github.com/vmatyusGitHub/node-sdk-core/blob/2269660441b93d426932ec9b0dfb51ac1dc0159a/auth/token-managers/token-manager.ts#L98
In the constructor `TokenInfo` is initialised to an empty object and `tokenName` shall be initialised in a subclass.
Are you suggesting to have only `accesToken` property with type string, this way I shall remove `tokenInfo` and `tokenName` from the object's property list?

Reply:
Since `tokenInfo` is used in the logic contained in this file, I'm fine with it existing here.

Reply:
Here was my thinking: This class requires its subclasses to implement `saveTokenInfo`, which is passed the response of the token request.  `saveTokenInfo` has to set some member variables ... specifically `expireTime`, `refreshTime`, and `tokenInfo`.

But what is `tokenInfo`?  It is just an object that must contain a field named `tokenName` that contains the access token.  From the point of view of a generic `TokenManager`, these are extraneous details.  All it needs is the `accessToken`.  So why not simplify the `TokenManager` class (i.e. make it more general), and simply use `accessToken` set by `saveTokenInfo`, and push the complexity of varying token property names into the subclass, `JWTTokenManager`.

I can live with the code the way it is, but I think it would be better if `tokenInfo` and `tokenName` was pushed down into `JWTTokenManager` and `TokenManager` only dealt with `accessToken`.

Reply:
Now, I understand your request. Thank you for the more detailed description. I created a modification that would use `accessToken`. In this commit the change can be reviewed: 2f5fe588dbe3d462e0ab98e323ffdc797f955f4f

Reply:
I was just wondering why it is important to use `TokenInfo` inside `JwtTokenManager`? Will there be a feature in the future that will be built upon the content of this object?

Reply:
Are you referring to the _type_ `TokenInfo`? I don't know that we have a specific feature in mind for the future so it is not truly important. However, I think it is worth leaving in as at the very least it provides a developer with knowledge of what they can expect the response to look like.